contents/javascript/Javascript-framework-Redux.md

```diff
@@ -0,0 +1,226 @@
+        if (index === action.index) {
+          return Object.assign({}, todo, {
+            completed: true
+          })
+        }
+        return todo
+      })
+    default:
+      return state
+  }
+}
+import { combineReducers, createStore } from 'redux'
+const reducer = combineReducers({ visibilityFilter, todos })
+const store = createStore(reducer)
+```
+
+## Why use Redux?
+
+([Adapted from this artcle](https://blog.logrocket.com/why-use-redux-reasons-with-clear-examples-d21bffd5835/))
+([And this article](https://www.smashingmagazine.com/2016/06/an-introduction-to-redux/))
+
+Now that we know what Redux is, let us take a look at some of its benefits that were made possible with these concepts of Redux. 
+
+1. **State is made predictable**<br/>
+The features of Redux make state predictable: There is always one source of truth in the store. This store is immutable. Reducers are pure functions, always producing the same result for the same state and action. And with these features, it is possible to implement functionality that is otherwise difficult, such as Undo and Redo. <br/><br/> Because neither the view, nor the network callbacks directly write to state. And changes to state via actions are centralized and happen in strict order, there is no need to worry for race conditions in our applications
+
+2. **Easier debug and inspect**<br/>
+Since actions and state are plain objects, it is possible to serialize them and log them. By logging them, it's easy to understand errors in code. You can also print the store within the code itself (for example using `console.log(store.getState())`{.js}). <br/><br/> The Redux DevTools is an amazing tool as well. Used in the browser, it provides features such as a tree visualization of the application state, as well a time travel ability to replay state changes. 
+
+![Redux Devtool](https://user-images.githubusercontent.com/7957859/48663602-3aac4900-ea9b-11e8-921f-97059cbb599c.png)
+
+
+<center>
+
+_Redux DevTools_ <sup>[image source](https://user-images.githubusercontent.com/7957859/48663602-3aac4900-ea9b-11e8-921f-97059cbb599c.png)</sup>
+</center>
+
+3. **Easier testing**<br/>
+It is easy make tests since the reducers used to change the state are pure functions. When writing code with Redux means writing small, pure and isolated reducer functions, we simply write small and isolated tests. Using Redux makes for testable code.
+
+4. **Server-side rendering**<br/>
+Redux can be used for server-side rendering, by sending the store created on the server side in response to the client's server request. Information from the store then can be used for the initial render of the application. 
+
+## Why not to use Redux?
+
+While Redux has its benefits, you shouldn't jump to Redux for every application you make.
+
+As Dan Abramov, one of the creators of Redux, says:
+
+<box light>
+I would like to amend this: don't use Redux until you have problems with vanilla React.
```

Comment:
Could remove this, and provide a brief framework independent explanation of why you shouldn't always use redux.
The user might mistake this for something to do with React, or might not know react.

Reply:
Noted.